### PR TITLE
1.0 pre3 release

### PR DESCRIPTION
**Pull request by o-smirnov**
@Athanaseus this one is ready to go, but we need to discuss how to make it properly installable from PyPI.

Firstly, I'm not sure how to install this outside of a virtualenv (and whether anyone would want to). At core this is a custom Jupyter kernel. I simply don't know how to install a custom Jupyter kernel without an associated virtualenv, or if this is even possible. 

If this is not possible (or we decide not to bother trying), we should at least make ``setup.py`` crash out if it detects a non-virtualenv install.

Secondly, after installing this into a virtualenv, there are some additional steps that need to be taken. These are currently done by ``bin/setup_radiopadre_virtualenv``. If you can just modify ``setup.py`` to invoke it after the install, that's probably fine.

### CODE REVIEW

**Comment by o-smirnov** [comment]
@gijzelaerr may have opinions on the above too.

**Comment by gijzelaerr** [comment]
It sounds like you need write permissions to the notebook folder? Terrible idea. Did one of you ever read in into the plugin infrastructure of notebooks? Those should probably facility such needs. I'm at the moment unfamiliar with this design nor I have the time to look into this.

**Comment by o-smirnov** [comment]
Well hey if you're not part of the solution, you're part of the problem. :P I know you don't have time now -- I was hoping for a ready answer. If you don't have one, get back to your honeymoon preparations. 

Yes I read about notebook plugins, that is a different beast entirely. But I was hoping someone younger and smarter would figure this out.

No, of course the user doesn't need write permission to the notebook folder. The installer does though. I have not yet discovered a way to install stuff without write permissions. There is web content that needs to be served up as part of the notebook (JS9 web pages and scripts, for example), how am I supposed to install that without write permissions?

**Comment by adrianjhpc** [comment]
Can you not just change the directory permissions in the installed, install the files, then change them back?

**Comment by o-smirnov** [comment]
Well, permissions are not the problem *yet* (@gijzelaerr is just getting his Python Police cap on and firing from the hip without understanding what I'm trying to do). Here's what needs to be installed:

* Binary components of JS9 need to be built and installed. This is annoying, as it it's old-school ``./configure && make install`` kind of thing.

* A bunch of web pages associated with JS9 etc. **must** go in under ``lib/python3.6/site-packages/notebook/static/``, because that's the only location Juputer will serve HTML components from.

* A custom kernel needs to be registered with ``ipython -m ipykernel install [--user] --name radiopadre``. In ``--user`` mode, this creates ``~/.local/share/jupyter/kernels/radiopadre``. In systemwide mode, I'm not sure.

* ``radiopadre-kernel.js`` needs to be copied or symlinked as ``kernel.js`` in the above kernel directory.

* A recent version of CARTA needs to be pulled down and untarred (the KERN one is too old).

So all of this is pretty straightforward to do in a virtualenv (with Jupyter itself living under the virtualenv, with its ``lib/python3.6/site-packages/notebook/static/`` writable). The ``setup-radiopadre-virtualenv`` script will do all these steps. 

It's how to map all this to a systemwide install that I'm not sure about, and whether a systemwide install is even worth the effort, given how easy it is for any given user to let it deploy to a virtualenv (or use the Docker image).

Note that none of this shit applies to ``radiopadre-client``. The client part has literally zero dependencies outside of Python itself.

**Comment by gijzelaerr** [comment]
sorry to sound like a grumpy old fart, but I still think it is a bad hack to dump stuff into the jupyter notebook serving folder, also from a security perspective (if you start fiddling with permissions). There must be a better way, but I just don't know (at the moment). So I'll just shut up from now on ;)

**Comment by ericmandel** [comment]
Can I help with this? What would be new school?

**Comment by o-smirnov** [comment]
> it is a bad hack to dump stuff into the jupyter notebook serving folder, also from a security perspective (if you start fiddling with permissions). 

We are not fiddling with permissions (fake news!). We just need to, like, you know, *install* stuff there at installation time.

And bad hack or not, it's the only way, by definition. Jupyter will actively cripple any HTML code *not* served from within its ``static`` folder (precisely due to security considerations), so custom code has to go there, or it won't run at all. 

I guess this argues for disabling system-wide install entirely? I'm ok with that -- see my original comment.

**Comment by o-smirnov** [comment]
> Can I help with this? What would be new school?

Thanks, @ericmandel! I wouldn't know though, I'm now officially on the old fart end of the spectrum myself. :)

At some point, @Athanaseus ought to take a look at the build system with an eye to incorporating JS9 in KERN as a standalone package, and work with you to get it done. 

This shouldn't be coupled to the present conversation though because (a) the way JS9 is bundled with radiopadre now serves the purpose, even if in a non-purist way, and (b) a JS9 KERN package wouldn't solve all of the installation problems above anyway (it's Javascript code would *still* need to go into Jupyter's static folder...)

**Comment by ericmandel** [comment]
Yes, I’d like to learn how the kids are doing things these days … though I’m not in a hurry to add new tasks to my Trello list!

So we’ll talk when the the time is right … (is this all in Docker? Would an image containing pre-built JS9 help?)

**Comment by o-smirnov** [comment]
> (is this all in Docker? Would an image containing pre-built JS9 help?)

No, we're setting this up as a "native" install. There is a Docker option as well, but that's the easy part. Anyway, from the commits, I can see that @Athanaseus is most of the way there.

File: bin/setup-radiopadre-virtualenv [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+            if not options.skip_git_pull:
+                message("Will do a git pull in {js9_dir}")
+                if shell("git -C {js9_dir} pull") != 0:
+                    js9status.write("git pull failed")
+                    bye("Failed to git pull in JS9 directory, see log above")
+
+    # Configure and install
+    js9_www = PADRE_VENV + "/js9-www"
+    if subprocess.call("""cd {js9_dir} && \
+            ./configure --prefix={PADRE_VENV} --with-webdir={js9_www} --with-helper=nod --with-cfitsio={options.cfitsio_path} && \
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: @o-smirnov is it supposed to be `--with-helper=nodejs`?

Comment (o-smirnov) [reply bin/setup-radiopadre-virtualenv]: Yes it ought to say ``nodejs``. Dunno how this built at all -- maybe @ericmandel is not so strict on checking the value....

**Pull request merged**